pallene/checker.lua

```diff
@@ -586,27 +612,50 @@ function FunChecker:check_var(var)
         elseif cname._tag == "checker.Name.Function" then
             var._type = self.p.module.functions[cname.id].typ
         elseif cname._tag == "checker.Name.Builtin" then
-            var._type = builtins[cname.name].typ
-               else
+            var._type = builtins.functions[cname.name].typ
+        elseif cname._tag == "checker.Name.Module" then
+            -- Module names can appear only in the dot notation.
+            -- For example, a statement like `local x = io` is illegal.
+            type_error(var.loc,
+                "cannot reference module name '%s' without dot notation",
+                var.name)
+        else
             error("impossible")
         end
 
     elseif tag == "ast.Var.Dot" then
-        var.exp = self:check_exp_synthesize(var.exp)
-        local ind_type = var.exp._type
-        if not types.is_indexable(ind_type) then
-            type_error(var.loc,
-                "trying to access a member of value of type '%s'",
-                types.tostring(ind_type))
-        end
-        local field_type = types.indices(ind_type)[var.name]
-        if not field_type then
-            type_error(var.loc,
-                "field '%s' not found in type '%s'",
-                var.name, types.tostring(ind_type))
+        if var.exp._tag == "ast.Exp.Var" and builtins.modules[var.exp.var.name] then
```

Comment:
This also needs to check if `var.exp.var` is a `ast.Var.Name`. It turns out that the example I had previously provided where `var.exp.var` is a `var.Bracket` works by chance because `builtins.modules[nil]` is `nil`. However, this code will do weird things if `var.exp.var` is an `ast.Var.Dot` because that case also has a `.name` field.